bookserver/crud.py

```diff
@@ -494,3 +580,189 @@ async def create_user_chapter_progress_entry(
+            )
+        if request_data.autogradable:
+            where_clause = where_clause & (
+                (Question.autograde == "unittest")
+                | Question.question_type.in_(auto_gradable_q)
+            )
+        query = select(Question).where(where_clause)
+
+        async with async_session() as session:
+            res = await session.execute(query)
+            rslogger.debug(f"{res=}")
+            questionlist = [q.name for q in res]
+
+    return questionlist
+
+
+async def fetch_assignment_question(assignment_name: str, question_name: str):
+
+    query = select(AssignmentQuestion).where(
+        (Assignment.name == assignment_name)
+        & (Assignment.id == AssignmentQuestion.assignment_id)
+        & (AssignmentQuestion.question_id == Question.id)
+        & (Question.name == question_name)
+    )
+
+    async with async_session() as session:
+        res = await session.execute(query)
+        rslogger.debug(f"{res=}")
+        return AssignmentQuestionValidator.from_orm(res.scalars().first())
+
+
+async def fetch_user_experiment(sid: str, ab_name: str) -> int:
+    query = (
+        select(UserExperiment.exp_group)
+        .where((UserExperiment.sid == sid) & (UserExperiment.experiment_id == ab_name))
+        .order_by(UserExperiment.id)
+    )
+    async with async_session() as session:
+        res = await session.execute(query)
+        r = res.scalars().first()
+        rslogger.debug(f"{r=}")
+        return r
+
+
+async def create_user_experiment_entry(sid: str, ab: str, group: int):
+    new_ue = UserExperiment(sid=sid, exp_group=group, experiment_id=ab)
+    async with async_session.begin() as session:
+        session.add(new_ue)
+    return UserExperimentValidator.from_orm(new_ue)
+
+
+async def fetch_viewed_questions(sid, questionlist) -> List[str]:
+    query = select(Useinfo).where(
+        (Useinfo.sid == sid) & (Useinfo.div_id.in_(questionlist))
+    )
+    async with async_session() as session:
+        res = await session.execute(query)
+        rslogger.debug(f"{res=}")
+        rlist = [row.div_id for row in res]
+    return rlist
+
+
+async def fetch_previous_selections(sid) -> List[str]:
+    query = select(SelectedQuestion).where(SelectedQuestion.sid == sid)
+    async with async_session() as session:
+        res = await session.execute(query)
+        rslogger.debug(f"{res=}")
+        return [row.selected_id for row in res.scalars().all()]
+
+
+async def fetch_timed_exam(
+    sid: str, exam_id: str, course_name: str
+) -> TimedExamValidator:
+    query = (
+        select(TimedExam)
+        .where(
+            (TimedExam.div_id == exam_id)
+            & (TimedExam.sid == sid)
+            & (TimedExam.course_name == course_name)
+        )
+        .order_by(TimedExam.id.desc())
+    )
+    async with async_session() as session:
+        res = await session.execute(query)
+        rslogger.debug(f"{res=}")
+        return TimedExamValidator.from_orm(res.scalars().first())
```

Comment:
Use `one_or_none`?

Reply:
I thought that you had this fixed so that .first() would work and the validator would be None in the case where there were no results.  In fact I'm pretty sure I tested and experimented with both approaches until I settled on this one as the working model.  Did that now change without me knowing?